engine/consensus/matching/engine.go

```diff
@@ -480,30 +486,27 @@ func (e *Engine) sealableResults() ([]*flow.IncorporatedResult, error) {
 RES_LOOP:
 	for _, incorporatedResult := range e.incorporatedResults.All() {
 
-		// if we have not received the block yet, we will just keep rechecking
-		// until the block has been received or the result has been purged
+		// not finding the block header for an incorporated result is a fatal
+		// implementation bug, as we only add results to the IncorporatedResults
+		// mempool, where _both_ the block that incorporates the result as well
+		// as the block the result pertains to are known
 		block, err := e.headersDB.ByBlockID(incorporatedResult.Result.BlockID)
-		if errors.Is(err, storage.ErrNotFound) {
-			log.Debug().Msg("skipping result with unknown block")
-			continue
-		}
 		if err != nil {
 			return nil, fmt.Errorf("could not retrieve block: %w", err)
 		}
 
-		// look for previous result in mempool and storage
+		// Retrieve parent result / skip if parent result still unknown:
+		// Before we store a result into the incorporatedResults mempool, we store it in resultsDB.
+		// I.e. resultsDB contains a superset of all results stored in the mempool. Hence, we only need
+		// to check resultsDB. Any result not in resultsDB cannot be in incorporatedResults mempool.
 		previousID := incorporatedResult.Result.PreviousResultID
-		previous, _ := e.incorporatedResults.ByResultID(previousID)
-		if previous == nil {
-			var err error
-			previous, err = e.resultsDB.ByID(previousID)
-			if errors.Is(err, storage.ErrNotFound) {
-				log.Debug().Msg("skipping sealable result with unknown previous result")
-				continue
-			}
-			if err != nil {
-				return nil, fmt.Errorf("could not get previous result: %w", err)
-			}
+		previous, err := e.resultsDB.ByID(previousID)
+		if errors.Is(err, storage.ErrNotFound) {
+			log.Debug().Msg("skipping sealable result with unknown previous result")
```

Comment:
```suggestion
			e.log.Debug().Msg("skipping sealable result with unknown previous result")
```

Better to add some context to debug this case, i.e block_id, height of the executed block, and parent result id

Reply:
thanks for spotting this Leo 🙇  There were a few more of this in matching Engine. Fixed them all